components/core/src/server/integrationapiimpl.ts

```diff
@@ -361,16 +402,35 @@ export class IntegrationApiImpl implements IntegrationApi {
     return new OprClientHandlerRegistration(handlerFn, this.changeHandlers);
   }
 
-  installCustomHandler(path: string, handler: CustomRequestHandler): void {
-    this.server.installCustomHandler(path, handler);
+  installOfferProducer(producer: OfferProducer): void {
+    this.host.installOfferProducer(producer);
   }
 
-  installOfferProducer(producer: OfferProducer): void {
-    this.server.installOfferProducer(producer);
+  getExpressRouter(): Router {
+    return this.router;
   }
 
-  getExpressServer(): Express {
-    return this.server.getExpressServer();
+  installCustomHandler(path: string, handler: CustomRequestHandler) {
+    if (this.host.isStarted()) {
```

Comment:
No - you can add an express endpoint whenever you want. This check is there to enforce the expected lifecycle of a TenantNode. We only want people installing new integrations during the setup phase. If an endpoint is added at any other time, it would have to mean that someone was adding an endpoint in a listener callback, or in some other weird out-of-band thread they built for themselves. That would always be a mistake, so this check makes it perfectly clear that we won't allow this kind of thing.

The kind of URL rewriting you're suggesting is kind of what we're doing right now. All requests to any tenant node are actually handled by a single endpoint (which then hands off processing to a Router).